lib/plugins/cors.js

```diff
@@ -73,66 +120,234 @@ function matchOrigin(req, origins) {
+
+        // if either 3.1 or 3.2 failed, return 400.
+        if (!matchedOrigin || !validHeaders) {
+            return next(new errs.BadRequestError(errMsg));
+        }
+
+        // 3.4) check that the requested resource and method are valid. couple
+        // ways we can do this. user passes in an option that can either be:
+        //
+        //  - a restify server instance:
+        //      in this scenario, we use the router to 'simulate' a routing
+        //      request, which will populate res.methods with the supported
+        //      methods for that resource. this is an easy way to
+        //      automatically get full support without having to implement
+        //      server.opts on every single cors resource.
+        //
+        //      on a successful 'route', the router will populate
+        //      res.methods with the list of supported methods, and also
+        //      likely return a 405, but that's expected. iif router returns
+        //      a 404, that means the resource doesn't even exist, so we
+        //      need to send a 400 back.
+        //
+        //  - a function, that given req/res, returns an array of methods
+        //      a user provided function that will return an array of supported
+        //      methods given a req and res object.
+        //
+        //  - a value of false
+        //      don't do anything, user will manually implement server.opts for
+        //      every single resource.
+
+        var findMethodsFn;
+
+        if (typeof preflightStrategy === 'function') {
+            findMethodsFn = preflightStrategy;
+        } else if (typeof preflightStrategy === 'object' &&
+                   preflightStrategy.router &&
+                   preflightStrategy.router.find) {
+
+            var server = preflightStrategy;
+            findMethodsFn = function findMethods() {
+                // anonymouns fn to avoid bind here
+                var args = [].slice.apply(arguments);
+                server.router.find.apply(server.router, args);
+            };
+        } else {
+            findMethodsFn = function noop(_, _2, cb) {
+                return cb();
+            };
+        }
+
+        // restify router has a sig of function found(err, route, context) {
+        // but we don't use route or context in this case. for simplicity
+        // sake, the user facing preflightStrategy should just return an
+        // array of methods as the second param.
+        findMethodsFn(req, res, function found(err, resMethods) {
+
+            if (err) {
+                if (err instanceof errs.ResourceNotFoundError ||
+                    err.restCode === 'ResourceNotFound') {
+                    return next(new errs.BadRequestError(errMsg));
+                }
+            }
+
+            // look for it first on res, then on the returned value, then
+            // finally fall back to empty array
+            var supportedMethods = res.methods || resMethods || [];
```

Comment:
`res.methods` shouldn't be necessary if we properly call the callback in our `findMethodsFn`